recursive-methods.rb

```diff
@@ -1,34 +1,94 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 def factorial(n)
-  puts "Not implemented."
+  if n < 1
+    return 1
+  else
+    return n * factorial(n-1)
+  end
 end
 
-def reverse(s)
-  puts "Not implemented."
+def reverse(str)
+  if str.length == 1
+    return str
+  else
+    return reverse(str[1..-1])+ str[0]
+  end
 end
 
 def reverse_inplace(s)
-  puts "Not implemented."
+  i = 0
+  j = s.length - 1
+  return inplace(s, i, j)
+end
+
+def inplace(s,i,j)
+  if i >= j
+    return
+  else
+    temp = s[i]
+    s[i] = s[j]
+    s[j] = temp
+    inplace(s,i+1,j-1)
+  end
 end
 
 def bunny(n)
-  puts "Not implemented."
+  if n <= 0
+    return 0
+  elsif n == 1
+    return 2
+  else
+    return 2 + bunny(n-1)
+  end
 end
 
 def nested(s)
-  puts "Not implemented."
+  if s.empty?
+    return true
+  elsif s.length  == 1
+    return false
+  elsif s[0] != s[-1]
+    return nested(s[1..-2])
+  else
+    return false
+  end
 end
 
+
+
 def search(array, value)
-  puts "Not implemented."
+  if array.empty?
+    return false
+  elsif array[0] == value
+    return true
+  else
+    return search(array[1..-1],value)
```

Comment:
In the scenario where the value is not in the array, or it is the last element in the array, you are currently creating n new arrays, one in each recursion call stack. This is adding unnecessary space complexity. Suggestion: similar to reversing the string in place, this one could benefit from having the next index at which the element needs to be examined passed in.